pfcp/udp/udp_test.go

```diff
@@ -9,57 +9,146 @@ import (
+	"github.com/wmnsk/go-pfcp/message"
 )
 
-const testPfcpClientPort = 12345
+var heartbeatRequestReceived bool
 
-func TestRun(t *testing.T) {
-	// Set SMF Node ID
+type Server struct {
+	addr *net.UDPAddr
+	Conn *net.UDPConn
+}
+
+func HandlePfcpHeartbeatRequestTest(msg *udp.Message) {
+	heartbeatRequestReceived = true
+}
+
+func Dispatch(msg *udp.Message) {
+	if msg.PfcpMessage == nil {
+		return
+	}
+	msgType := msg.PfcpMessage.MessageType()
+	switch msgType {
+	case message.MsgTypeHeartbeatRequest:
+		HandlePfcpHeartbeatRequestTest(msg)
+	}
+}
+
+func (s *Server) Start() error {
+	conn, err := net.ListenUDP("udp", s.addr)
+	if err != nil {
+		return err
+	}
+	s.Conn = conn
+	return nil
+}
+
+func (s *Server) SendPFCPMessage(msg message.Message, remoteAddress *net.UDPAddr) error {
+	buf := make([]byte, msg.MarshalLen())
+	err := msg.MarshalTo(buf)
+	if err != nil {
+		return err
+	}
+
+	_, err = s.Conn.WriteToUDP(buf, remoteAddress)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-	context.SMF_Self().CPNodeID = pfcpType.NodeID{
-		NodeIdType:  pfcpType.NodeIdTypeIpv4Address,
+func TestRun(t *testing.T) {
+	context.SMF_Self().CPNodeID = context.NodeID{
+		NodeIdType:  context.NodeIdTypeIpv4Address,
 		NodeIdValue: net.ParseIP("127.0.0.1").To4(),
 	}
+	context.SMF_Self().PFCPPort = 8811
 
-	udp.Run(smf_pfcp.Dispatch)
-
-	testPfcpReq := pfcp.Message{
-		Header: pfcp.Header{
-			Version:         1,
-			MP:              0,
-			S:               0,
-			MessageType:     pfcp.PFCP_ASSOCIATION_SETUP_REQUEST,
-			MessageLength:   9,
-			SEID:            0,
-			SequenceNumber:  1,
-			MessagePriority: 0,
-		},
-		Body: pfcp.PFCPAssociationSetupRequest{
-			NodeID: &pfcpType.NodeID{
-				NodeIdType:  0,
-				NodeIdValue: net.ParseIP("192.168.1.1").To4(),
-			},
-		},
-	}
-
-	srcAddr := &net.UDPAddr{
+	localAddr := &net.UDPAddr{
 		IP:   net.ParseIP("127.0.0.1"),
-		Port: testPfcpClientPort,
+		Port: 8811,
 	}
-	dstAddr := &net.UDPAddr{
+	remoteAddr := &net.UDPAddr{
 		IP:   net.ParseIP("127.0.0.1"),
-		Port: pfcpUdp.PFCP_PORT,
+		Port: 1234,
+	}
+	go udp.Run(Dispatch)
+	err := udp.WaitForServer()
+	if err != nil {
+		t.Fatalf("Failed to start PFCP server: %v", err)
 	}
 
-	err := pfcpUdp.SendPfcpMessage(testPfcpReq, srcAddr, dstAddr)
-	require.Nil(t, err)
+	if udp.Server == nil {
+		t.Fatalf("Expected Server to be initialized")
+	}
+
+	if udp.Server.Conn == nil {
+		t.Fatalf("Expected Server to be listening")
+	}
+	defer udp.Server.Conn.Close()
+
+	setupRequest := message.NewHeartbeatRequest(
+		1,
+		ie.NewRecoveryTimeStamp(time.Now()),
+		nil,
+	)
+
+	server := &Server{
+		addr: remoteAddr,
+	}
+	server.Start()
+	server.SendPFCPMessage(setupRequest, localAddr)
 
-	time.Sleep(300 * time.Millisecond)
+	time.Sleep(1 * time.Second)
+
+	if !heartbeatRequestReceived {
+		t.Error("Expected Heartbeat Request to be received")
+	}
+}
+
+func TestServerSendPfcp(t *testing.T) {
+	localAddress := &net.UDPAddr{
+		IP:   net.ParseIP("127.0.0.1"),
+		Port: 8805,
```

Comment:
What about using the values from `factory/config.go`?
```suggestion
		Port: factory.DEFAULT_PFCP_PORT,
```

Reply:
Done